tokio/src/process/unix/pidfd_reaper.rs

```diff
@@ -0,0 +1,310 @@
+    process::{
+        imp::{orphan::Wait, OrphanQueue},
+        kill::Kill,
+    },
+    util::error::RUNTIME_SHUTTING_DOWN_ERROR,
+};
+
+use libc::{syscall, SYS_pidfd_open, __errno_location, ENOSYS, PIDFD_NONBLOCK};
+use mio::{event::Source, unix::SourceFd};
+use std::{
+    fs::File,
+    future::Future,
+    io,
+    marker::Unpin,
+    ops::Deref,
+    os::unix::io::{AsRawFd, FromRawFd, RawFd},
+    pin::Pin,
+    process::ExitStatus,
+    sync::atomic::{AtomicBool, Ordering::Relaxed},
+    task::{Context, Poll},
+};
+
+#[derive(Debug)]
+struct Pidfd {
+    fd: File,
+}
+
+impl Pidfd {
+    fn open(pid: u32) -> Option<Pidfd> {
+        // Store false (0) to reduce executable size
+        static NO_PIDFD_SUPPORT: AtomicBool = AtomicBool::new(false);
+
+        if NO_PIDFD_SUPPORT.load(Relaxed) {
+            return None;
+        }
+
+        unsafe {
+            let fd = syscall(SYS_pidfd_open, pid, PIDFD_NONBLOCK);
+            if fd == -1 {
+                let errno = *__errno_location();
+
+                if errno == ENOSYS {
+                    NO_PIDFD_SUPPORT.store(true, Relaxed)
+                }
+
+                None
+            } else {
+                Some(Pidfd {
+                    fd: File::from_raw_fd(fd as i32),
+                })
+            }
+        }
+    }
+}
+
+impl AsRawFd for Pidfd {
+    fn as_raw_fd(&self) -> RawFd {
+        self.fd.as_raw_fd()
+    }
+}
+
+impl Source for Pidfd {
+    fn register(
+        &mut self,
+        registry: &mio::Registry,
+        token: mio::Token,
+        interest: mio::Interest,
+    ) -> io::Result<()> {
+        SourceFd(&self.as_raw_fd()).register(registry, token, interest)
+    }
+
+    fn reregister(
+        &mut self,
+        registry: &mio::Registry,
+        token: mio::Token,
+        interest: mio::Interest,
+    ) -> io::Result<()> {
+        SourceFd(&self.as_raw_fd()).reregister(registry, token, interest)
+    }
+
+    fn deregister(&mut self, registry: &mio::Registry) -> io::Result<()> {
+        SourceFd(&self.as_raw_fd()).deregister(registry)
+    }
+}
+
+#[derive(Debug)]
+struct PidfdReaperInner<W>
+where
+    W: Unpin,
+{
+    inner: W,
+    pidfd: PollEvented<Pidfd>,
+}
+
+#[allow(deprecated)]
+fn is_rt_shutdown_err(err: &io::Error) -> bool {
+    if let Some(inner) = err.get_ref() {
+        // Using `Error::description()` is more efficient than `format!("{inner}")`,
+        // so we use it here even if it is deprecated.
+        inner.source().is_none() && inner.description() == RUNTIME_SHUTTING_DOWN_ERROR
+    } else {
+        false
+    }
+}
+
+impl<W> Future for PidfdReaperInner<W>
+where
+    W: Wait + Unpin,
+{
+    type Output = io::Result<ExitStatus>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = Pin::into_inner(self);
+
+        match ready!(this.pidfd.poll_read_ready(cx)) {
+            Err(err) if err.kind() == io::ErrorKind::Other && is_rt_shutdown_err(&err) => {
```

Comment:
Can we move `err.kind() == io::ErrorKind::Other` into the function call?

Reply:
Thanks, I've moved it into `is_rt_hust down_err(_)`